patches/minecraft/net/minecraft/data/AdvancementProvider.java.patch

```diff
@@ -0,0 +1,44 @@
+--- a/net/minecraft/data/AdvancementProvider.java
++++ b/net/minecraft/data/AdvancementProvider.java
+@@ -24,11 +24,15 @@
+    private static final Gson field_204024_b = (new GsonBuilder()).setPrettyPrinting().create();
+    private final DataGenerator field_204025_c;
+    private final List<Consumer<Consumer<Advancement>>> field_204283_d = ImmutableList.of(new EndAdvancements(), new HusbandryAdvancements(), new AdventureAdvancements(), new NetherAdvancements(), new StoryAdvancements());
++   private net.minecraftforge.common.data.ExistingFileHelper fileHelper;
+ 
++   @Deprecated
+    public AdvancementProvider(DataGenerator p_i48869_1_) {
+       this.field_204025_c = p_i48869_1_;
+    }
+ 
++   public AdvancementProvider(DataGenerator generatorIn, net.minecraftforge.common.data.ExistingFileHelper fileHelperIn){ this.field_204025_c = generatorIn; this.fileHelper = fileHelperIn;}
++
+    public void func_200398_a(DirectoryCache p_200398_1_) throws IOException {
+       Path path = this.field_204025_c.func_200391_b();
+       Set<ResourceLocation> set = Sets.newHashSet();
+@@ -47,12 +51,25 @@
+          }
+       };
+ 
++      registerAdvancement(consumer, fileHelper);
++
++      if(this.getClass() == AdvancementProvider.class)//Forge: use registerAdvancement instead.
+       for(Consumer<Consumer<Advancement>> consumer1 : this.field_204283_d) {
+          consumer1.accept(consumer);
+       }
+ 
+    }
+ 
++   /**
++    * Subclasses should override this function if they want to generate Advancements.
++    * Just use {@see Advancement.Builder} to build your Advancements, you don't need an extra consumer like the vanilla classes.
++    * @param consumer used for the build function from {@see Advancement.Builder}
++    * @param fileHelper used for the build function from {@see Advancement.Builder}
++    */
++   protected void registerAdvancement(Consumer<Advancement> consumer, net.minecraftforge.common.data.ExistingFileHelper fileHelper){
```

Comment:
I would suggest maybe a rename to `registerAdvancements`, to be consistent with the other similarly-functioning methods in datagen? (`registerTags` for `TagsProvider`, `registerRecipes` in `RecipeProvider`, `registerStatesAndModels` in `BlockStateProvider`, etc.)

Also, a suggestion for the javadoc:
```java
/**
 * Override this method for registering and generating custom {@link Advancement}s.
 * @param consumer Consumer for registering advancements to this provider
 * @param fileHelper Existing file helper
 * @see Advancement.Builder#build(Consumer<Advancement>, ResourceLocation, net.minecraftforge.common.data.ExistingFileHelper)
 */
```

Reply:
i would like to keep the second part of my javadoc, since it is more clear what you should use, because if someone looks at the vanilla Classes they could be confused with the additional  consumer they use. Also Intellij does not show the last line when you look at the javadoc on the overridden Method.